### PR TITLE
Wave2 988 page manager custom layout options

### PR DESCRIPTION
**Pull request by robbdavis**

### CODE REVIEW

File: sites/all/modules/ucalgary/ucgy_core/ucgy_core_panels_frame/includes/stack-admin.inc [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+/**
+ * Submission handler for "delete"-ing a frame.
+ */
+function ucgy_panels_frame_stack_frame_delete_submit(&$form, &$form_state) {
+  $cache_key = $form_state['cache_key'];
+  $name = $form_state['frame'];
+
+  $item = $form_state['item'];
+  unset($item->data[$name]);
+  ucgy_panels_frame_cache_set('stack', $cache_key, $item);
```
Comment (gsowrey) [inline sites/all/modules/ucalgary/ucgy_core/ucgy_core_panels_frame/includes/stack-admin.inc]: What happens if $cache_key is blank? Invalid?

File: sites/all/modules/ucalgary/ucgy_core/ucgy_core_panels_frame/includes/stack-admin.inc [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+
+/**
+ * Validation handler to prevent nested layouts.
+ * @fixme Need a proper helper that builds layout name (ex: ucgy_panels_frame:stack:name).
+ */
+function ucgy_panels_frame_stack_frame_layout_validate(&$form, &$form_state) {
+  $name = $form_state['item']->name;
+  $frame = $form_state['values']['identifier'];
+  $layout = $form_state['values']['layout'];
+  if ($layout == 'ucgy_panels_frame_stack:' . $name) {
```
Comment (gsowrey) [inline sites/all/modules/ucalgary/ucgy_core/ucgy_core_panels_frame/includes/stack-admin.inc]: If $layout is unset?

File: sites/all/modules/ucalgary/ucgy_core/ucgy_core_panels_frame/includes/stack-admin.inc [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+
+/**
+ * Common submission handler for add/edit/clone operations.
+ */
+function ucgy_panels_frame_stack_frame_common_submit(&$form, &$form_state) {
+  $cache_key = $form_state['cache_key'];
+  $item = $form_state['item'];
+
+  foreach (array('label', 'identifier', 'weight', 'layout') as $key) {
+    $frame[$key] = $form_state['values'][$key];
```
Comment (gsowrey) [inline sites/all/modules/ucalgary/ucgy_core/ucgy_core_panels_frame/includes/stack-admin.inc]: What if $form_state['values'][$key] doesn't exist?

File: sites/all/modules/ucalgary/ucgy_core/ucgy_core_panels_frame/includes/stack-admin.inc [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+  $item = $form_state['item'];
+
+  // Wrapper to allow for ajax replacement.
+  $form['#prefix'] = '<div id="ucgy-panels-frame-stack-edit-form">';
+  $form['#suffix'] = '</div>';
+
+  // Call out the values that will have no UI here. It will be referenced in
+  // multiple places.
+  $no_ui = array('label', 'identifier', 'layout');
+  $cache_key = ucgy_panels_frame_cache_key($item->name);
```
Comment (gsowrey) [inline sites/all/modules/ucalgary/ucgy_core/ucgy_core_panels_frame/includes/stack-admin.inc]: Make sure $item is an object, and that $item->name exists

File: sites/all/modules/ucalgary/ucgy_core/ucgy_core_panels_frame/includes/stack-admin.inc [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+  foreach ($item->data as $frame_id => $frame) {
+    foreach ($no_ui as $hidden) {
+      $form['data'][$frame_id][$hidden] = array(
+        '#type' => 'value',
+        '#value' => $frame[$hidden],
+      );
+    }
+
+    // Icon
+    $layout = panels_get_layout($frame['layout']);
```
Comment (gsowrey) [inline sites/all/modules/ucalgary/ucgy_core/ucgy_core_panels_frame/includes/stack-admin.inc]: Check if $frame['layout'] exists

File: sites/all/modules/ucalgary/ucgy_core/ucgy_core_panels_frame/includes/stack-admin.inc [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+    // Loop through each property.
+    foreach (element_children($element[$frame]) as $property) {
+      // Filter out the properties that have no UI.
+      if (!in_array($property, $element['#no_ui'])) {
+        $rows[] = array('data' => render($element[$frame][$property]), 'class' => $property);
+      }
+    }
+    $table['rows'][] = array(
+      'class' => array('draggable'),
+      'data' => $rows,
```
Comment (gsowrey) [inline sites/all/modules/ucalgary/ucgy_core/ucgy_core_panels_frame/includes/stack-admin.inc]: if(isset($rows))...

File: sites/all/modules/ucalgary/ucgy_core/ucgy_core_panels_frame/css/ucgy-panels-frame.ui-stack.css [hunk 7]
```diff
@@ -0,0 +1,18 @@
+.ucgy-panels-frame-stack-ui-table td.icon .layout-icon {
+  float:left;
+}
+
+.stack-admin-info {
```
Comment (robbdavis) [inline sites/all/modules/ucalgary/ucgy_core/ucgy_core_panels_frame/css/ucgy-panels-frame.ui-stack.css]: Fixed.

File: sites/all/modules/ucalgary/ucgy_core/ucgy_core_panels_frame/includes/stack-admin.inc [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+  if ($form_state['executed']) {
+    $commands = array();
+
+    // Rebuild stack edit form with new settings.
+    $form = drupal_build_form('ucgy_panels_frame_stack_edit_form', $form_state);
+    $commands[] = ajax_command_replace('#ucgy-panels-frame-stack-edit-form', drupal_render($form));
+    $commands[] = ctools_modal_command_dismiss();
+  }
+
+  print ajax_render($commands);
```
Comment (robbdavis) [inline sites/all/modules/ucalgary/ucgy_core/ucgy_core_panels_frame/includes/stack-admin.inc]: Put in a check for $commands.

File: sites/all/modules/ucalgary/ucgy_core/ucgy_core_panels_frame/includes/stack-admin.inc (hunk 1 above)
Comment (robbdavis) [inline sites/all/modules/ucalgary/ucgy_core/ucgy_core_panels_frame/includes/stack-admin.inc]: The code in this file is from the original module (I never wrote it). I don't see any errors from this in the log. Is it better to wrap the function in 165 with an if(isset($cache_key) or let it degrade without that parameter?

File: sites/all/modules/ucalgary/ucgy_core/ucgy_core_panels_frame/includes/stack-admin.inc (hunk 2 above)
Comment (robbdavis) [inline sites/all/modules/ucalgary/ucgy_core/ucgy_core_panels_frame/includes/stack-admin.inc]: Initial layout is set in code so there will always be a layout.

File: sites/all/modules/ucalgary/ucgy_core/ucgy_core_panels_frame/includes/stack-admin.inc (hunk 3 above)
Comment (robbdavis) [inline sites/all/modules/ucalgary/ucgy_core/ucgy_core_panels_frame/includes/stack-admin.inc]: I'm not sure. I never wrote the code in this admin.inc. It's possible that this form state will always exist. I started putting in checks and ended up breaking something. I'm not seeing any errors in the log for this. How should I proceed.. it seems there are a lot of variables that don't have checks; I'm feeling like I might end up doing more damage than good.

File: sites/all/modules/ucalgary/ucgy_core/ucgy_core_panels_frame/includes/stack-admin.inc (hunk 5 above)
Comment (robbdavis) [inline sites/all/modules/ucalgary/ucgy_core/ucgy_core_panels_frame/includes/stack-admin.inc]: The frame layout will always exist.

File: sites/all/modules/ucalgary/ucgy_core/ucgy_core_panels_frame/includes/stack-admin.inc (hunk 4 above)
Comment (robbdavis) [inline sites/all/modules/ucalgary/ucgy_core/ucgy_core_panels_frame/includes/stack-admin.inc]: So if $item->name does not exist don't set $cache_key? Will $cache_key get set even if there is no parameter for ucgy_panels_frame_cache_key? if $cache key doesn't get set will that cause issues in lines 240 - 266?  Again, not having written the code and knowing exactly what's happening, I'm wondering if putting in these checks makes sense. (code is working and there are no errors from this in the log)

File: sites/all/modules/ucalgary/ucgy_core/ucgy_core_panels_frame/includes/stack-admin.inc (hunk 6 above)
Comment (robbdavis) [inline sites/all/modules/ucalgary/ucgy_core/ucgy_core_panels_frame/includes/stack-admin.inc]: $rows is defined as an array above.

File: sites/all/modules/ucalgary/ucgy_core/ucgy_core_panels_frame/css/ucgy-panels-frame.choose-layout.css [hunk 9]
```diff
@@ -0,0 +1,7 @@
+.ucgy-panels-frame-choose-layout .form-type-radio {
```
Comment (robbdavis) [inline sites/all/modules/ucalgary/ucgy_core/ucgy_core_panels_frame/css/ucgy-panels-frame.choose-layout.css]: It's specific to selecting layout for each stack.

**Comment by taeram** [comment]
Closing this PR on GitHub, but keeping the branch around for historical purposes.

**Pull request closed without merging**